cmake/FlagsSummary.cmake

```diff
@@ -0,0 +1,83 @@
+include_guard(GLOBAL)
+
+function(indent_message header content indent_num)
+  if(indent_num GREATER 0)
+    string(REPEAT " " ${indent_num} indentation)
+    string(REPEAT "." ${indent_num} tail)
+    string(REGEX REPLACE "${tail}$" "" header "${header}")
+  endif()
+  message("${indentation}${header} ${content}")
+endfunction()
+
+# Print tools' flags on best-effort. Include the abstracted
+# CMake flags that we touch ourselves.
+function(print_flags_per_config config indent_num)
+  string(TOUPPER "${config}" config_uppercase)
+
+  string(STRIP "${CMAKE_C_FLAGS} ${CMAKE_C_FLAGS_${config_uppercase}}" c_language_flags)
+  string(STRIP "${c_language_flags} ${CMAKE_C${CMAKE_C_STANDARD}_STANDARD_COMPILE_OPTION}" c_compiler_flags)
+  get_target_property(pic secp256k1 POSITION_INDEPENDENT_CODE)
+  if(pic AND CMAKE_C_COMPILE_OPTIONS_PIC)
+    string(APPEND c_compiler_flags " ${CMAKE_C_COMPILE_OPTIONS_PIC}")
+  endif()
+  if(CMAKE_C_COMPILE_OPTIONS_VISIBILITY AND CMAKE_C_VISIBILITY_PRESET)
+    string(APPEND c_compiler_flags " ${CMAKE_C_COMPILE_OPTIONS_VISIBILITY}${CMAKE_C_VISIBILITY_PRESET}")
+  endif()
+  get_directory_property(compile_options COMPILE_OPTIONS)
+  list(JOIN compile_options " " compile_options)
+  string(STRIP "${c_compiler_flags} ${compile_options}" c_compiler_flags)
+  string(STRIP "${c_compiler_flags} ${SECP256K1_APPEND_CFLAGS}" c_compiler_flags)
+  indent_message("C compiler flags ......................" "${c_compiler_flags}" ${indent_num})
+
+  if(BUILD_SHARED_LIBS)
+    string(STRIP "${CMAKE_SHARED_LINKER_FLAGS} ${CMAKE_SHARED_LINKER_FLAGS_${config_uppercase}}" linker_flags)
+    if(NOT APPLE AND NOT MSVC)
+      string(STRIP "${CMAKE_SHARED_LIBRARY_C_FLAGS} ${linker_flags}" linker_flags)
+    endif()
+    if(NOT MSVC)
+      string(STRIP "${c_language_flags} ${linker_flags}" linker_flags)
+    endif()
+    if(APPLE)
+      get_target_property(compatibility_version secp256k1 MACHO_COMPATIBILITY_VERSION)
+      if(compatibility_version)
+        string(STRIP "${linker_flags} ${CMAKE_C_OSX_COMPATIBILITY_VERSION_FLAG}${compatibility_version}" linker_flags)
+      endif()
+      get_target_property(current_version secp256k1 MACHO_CURRENT_VERSION)
+      if(compatibility_version)
+        string(STRIP "${linker_flags} ${CMAKE_C_OSX_CURRENT_VERSION_FLAG}${current_version}" linker_flags)
+      endif()
+    endif()
+    indent_message("Linker flags .........................." "${linker_flags}" ${indent_num})
+  else()
+    string(STRIP "${CMAKE_STATIC_LINKER_FLAGS} ${CMAKE_STATIC_LINKER_FLAGS_${config_uppercase}}" archiver_options)
+    indent_message("Archiver options ......................" "${archiver_options}" ${indent_num})
```

Comment:
This doesn't seem to list the options passed to the archiver? i.e:
```bash
cmake -B build -DCMAKE_C_COMPILER=clang -DBUILD_SHARED_LIBS=OFF
...
Build artifacts:
  library type ........................ Static
...
C compiler ............................ Clang 17.0.6, /usr/bin/clang
CMAKE_BUILD_TYPE ...................... RelWithDebInfo
C compiler flags ...................... -O2 -g -std=c90 -fvisibility=hidden -pedantic -Wall -Wcast-align -Wconditional-uninitialized -Wextra -Wnested-externs -Wno-long-long -Wno-overlength-strings -Wno-unused-function -Wreserved-identifier -Wshadow -Wstrict-prototypes -Wundef
Archiver options ...................... 
....
/usr/bin/llvm-ar qc libsecp256k1.a CMakeFiles/secp256k1.dir/secp256k1.c.o CMakeFiles/secp256k1_precomputed.dir/precomputed_ecmult.c.o CMakeFiles/secp256k1_precomputed.dir/precomputed_ecmult_gen.c.o
/usr/bin/llvm-ranlib libsecp256k1.a
```

Shouldn't this be listing `qc` as the archiver options? Otherwise, what is the output expected to be there?

Reply:
> Shouldn't this be listing `qc` as the archiver options?

Isn't `qc` rather a operation mode, not options?

>  Otherwise, what is the output expected to be there?

Archiver options are expected, For instance,
```
$ cmake -B build -DBUILD_SHARED_LIBS=OFF -DCMAKE_STATIC_LINKER_FLAGS=-o
...
Archiver options ...................... -o
```

Reply:
Thanks! [Reworked](https://github.com/bitcoin-core/secp256k1/pull/1558#issuecomment-2200956197).